tracer/src/Datadog.Trace/OTelMetrics/MetricReaderHandler.cs

```diff
@@ -0,0 +1,202 @@
+// <copyright file="MetricReaderHandler.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#if NET6_0_OR_GREATER
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.OTelMetrics
+{
+    internal static class MetricReaderHandler
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(MetricReaderHandler));
+
+        // Temporary storage for aggregation (like RuntimeMetrics._exceptionCounts)
+        // Will be cleared after each export to prevent memory leaks
+        private static readonly ConcurrentDictionary<string, MetricPoint> CapturedMetrics = new();
+
+        private static readonly Dictionary<string, object?> EmptyTags = new();
+        private static string? _cachedTemporality;
+
+        public static void OnInstrumentPublished(Instrument instrument, MeterListener listener)
+        {
+            bool shouldEnable;
+            var meterName = instrument.Meter.Name;
+            var enabledMeterNames = Tracer.Instance.Settings.OpenTelemetryMeterNames;
+
+            if (enabledMeterNames.Length > 0)
+            {
+                shouldEnable = enabledMeterNames.Contains(meterName);
+            }
+            else
+            {
+                shouldEnable = !meterName.StartsWith("System.", StringComparison.Ordinal);
+            }
+
+            if (shouldEnable)
+            {
+                listener.EnableMeasurementEvents(instrument, state: null);
+                Log.Debug("Enabled measurement events for instrument: {InstrumentName} from meter: {MeterName}", instrument.Name, meterName);
+            }
+        }
+
+        public static void OnMeasurementRecordedLong(Instrument instrument, long value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            ProcessMeasurement(instrument, value, tags, isInteger: true);
+        }
+
+        public static void OnMeasurementRecordedDouble(Instrument instrument, double value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            ProcessMeasurement(instrument, value, tags, isInteger: false);
+        }
+
+        private static void ProcessMeasurement<T>(Instrument instrument, T value, ReadOnlySpan<KeyValuePair<string, object?>> tags, bool isInteger)
+            where T : struct
+        {
+            try
+            {
+                var instrumentType = instrument.GetType().FullName;
```

Comment:
Using my suggestion from [this comment](https://github.com/DataDog/dd-trace-dotnet/pull/7511/files#r2350327923) to have a `state` object for each Instrument, you will then have a state object when measurement callback is made. I recommend determining the "instrument type" once during the state creation so you don't have to recalculate it every single time a new measurement is made.